utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,601 @@
+
+textureDescriptorTypeToText :: TextureDescriptorType -> Text
+textureDescriptorTypeToText = \case
+  Sampler2D -> "sampler2D"
+
+instance FromJSON TextureDescriptorType where
+  parseJSON = withTextMaybe "type" textureDescriptorTypeFromText
+
+instance ToJSON TextureDescriptorType where
+  toJSON     = String . textureDescriptorTypeToText
+  toEncoding = toEncoding . textureDescriptorTypeToText
+
+convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType
+convertTextureDescriptorType = \case
+  Sampler2D -> DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER
+
+makeTextureDescriptorSetLayoutBinding
+  :: ShaderStage -> Texture -> (Int, DescriptorSetLayoutBinding)
+makeTextureDescriptorSetLayoutBinding stage Texture {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = convertTextureDescriptorType type'
+         , descriptorCount = maybe 1 (fromIntegral . V.sum) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+makeDescriptorSetLayoutBindings
+  :: ShaderStage
+  -> Vector Ubo
+  -> Vector Texture
+  -> Vector (Int, DescriptorSetLayoutBinding)
+makeDescriptorSetLayoutBindings stage ubos textures =
+  uboBindings <> textureBindings
+ where
+  uboBindings = V.map (makeUboDescriptorSetLayoutBinding stage) ubos
+  textureBindings =
+    V.map (makeTextureDescriptorSetLayoutBinding stage) textures
+
+makeDescriptorSetLayoutCreateInfos
+  :: Vector (Int, DescriptorSetLayoutBinding)
+  -> V.Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorSetLayoutCreateInfos bindings =
+  V.map makeDescriptorSetLayoutCreateInfo
+    . V.fromList
+    . M.elems
+    . M.unionWith (V.++) sets
+    $ setsMap
+ where
+  setSize = V.maximum . (fst <$>) $ bindings :: Int
+  sets :: Map Int (Vector DescriptorSetLayoutBinding)
+  sets = M.fromList . map (, []) $ [0 .. setSize]
+  setsMap :: Map Int (Vector DescriptorSetLayoutBinding)
+  setsMap =
+    M.fromList
+      . map (liftA2 (,) (fst . head) (V.fromList . (snd <$>)))
+      . groupBy ((==) `on` fst)
+      . sortOn fst
+      . V.toList
+      $ bindings
+
+makeDescriptorSetLayoutCreateInfo
+  :: Vector DescriptorSetLayoutBinding -> DescriptorSetLayoutCreateInfo '[]
+makeDescriptorSetLayoutCreateInfo bindings = zero { bindings = bindings }
+
+data VertexAttributeType
+  = Vec2
+  | Vec3
+  | Vec4
+  deriving (Show)
+
+vertexAttributeTypeFromText :: Text -> Maybe VertexAttributeType
+vertexAttributeTypeFromText = \case
+  "vec2" -> Just Vec2
+  "vec3" -> Just Vec3
+  "vec4" -> Just Vec4
+  _      -> Nothing
+
+vertexAttributeTypeToText :: VertexAttributeType -> Text
+vertexAttributeTypeToText = \case
+  Vec2 -> "vec2"
+  Vec3 -> "vec3"
+  Vec4 -> "vec4"
+
+instance FromJSON VertexAttributeType where
+  parseJSON = withTextMaybe "type" vertexAttributeTypeFromText
+
+instance ToJSON VertexAttributeType where
+  toJSON     = String . vertexAttributeTypeToText
+  toEncoding = toEncoding . vertexAttributeTypeToText
```

Comment:
runaway toEncoding...

Reply:
I use a `traceShow` to trace the call and found it won't cause a dead loop.

`toEncoding = traceShow "run here" . toEncoding @Text . to`

and here is a example to explain it, run `foo (2::Int)` will return a `1`
``` haskell
import Text (Text)
import qualified Text as T

class Foo a where
  foo :: a -> Int

instance Foo Int where
  foo = foo @Text . bar

instance Foo Text where
  foo _ = 1

bar :: Int -> Text
bar = T.pack . show
```

Reply:
Sorry, I've meant there are more places like this, defining `toEncoding` manually where `toJSON` would suffice.

Reply:
manually implemented `toEncoding` are removed